src/librustc/ty/context.rs

```diff
@@ -1687,13 +1660,49 @@ pub mod tls {
         TLV.with(|tlv| tlv.get())
     }
 
+    impl<'tcx> TlsContext for TyCtxt<'tcx> {
```

Comment:
I'd rename this to `QueryContext` and keep `ImplicitCtxt` in `rustc`. Instead use 2 methods to read/write to `ImplicitCtxt` like `DepContext`.